File: bindings/java/src/main/java/org/indunet/robot/bus/NativeLoader.java
package org.indunet.robot.bus;

import com.sun.jna.Native;
import com.sun.jna.Platform;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Resolves the shared C ABI library built from {@code bindings/cpp/native}
 * ({@code librobot_bus_c} / {@code robot_bus_c}).
 *
 * <p>Search order:
 *
 * <ol>
 *   <li>Already loaded via {@link #markPreloaded()} (Android {@code RobotBusAndroid.init})
 *   <li>{@code ROBOT_BUS_NATIVE} / {@code -Drobot.bus.native} — full path to the shared library
 *   <li>{@code ROBOT_BUS_NATIVE_DIR} / {@code -Drobot.bus.native.dir} — directory containing it
 *   <li>Sibling release build: {@code bindings/cpp/native/target/release}
 *   <li>System / APK library name {@code robot_bus_c}
 * </ol>
 */
final class NativeLoader {
    private static volatile boolean preloaded = false;

    private NativeLoader() {}

    static void markPreloaded() {
        preloaded = true;
    }

    static RobotBusC loadLibrary() {
        if (preloaded || isAndroid()) {
            return Native.load("robot_bus_c", RobotBusC.class);
        }
        String explicit = System.getenv("ROBOT_BUS_NATIVE");
        if (explicit == null || explicit.isBlank()) {
            explicit = System.getProperty("robot.bus.native");
        }
        if (explicit != null && !explicit.isBlank()) {
            return Native.load(explicit, RobotBusC.class);
        }
        String dir = System.getenv("ROBOT_BUS_NATIVE_DIR");
        if (dir == null || dir.isBlank()) {
            dir = System.getProperty("robot.bus.native.dir");
        }
        List<Path> candidates = new ArrayList<>();
        if (dir != null && !dir.isBlank()) {
            candidates.add(Path.of(dir));
        }
        candidates.add(Path.of("").toAbsolutePath().resolve("../cpp/native/target/release").normalize());
        candidates.add(Path.of(System.getProperty("user.dir"), "bindings/cpp/native/target/release"));
        candidates.add(Path.of(System.getProperty("user.dir"), "../cpp/native/target/release").normalize());
        for (Path candidate : candidates) {
            Path file = findInDir(candidate);
            if (file != null) {
                return Native.load(file.toAbsolutePath().toString(), RobotBusC.class);
            }
        }
        return Native.load("robot_bus_c", RobotBusC.class);
    }

    private static boolean isAndroid() {
        try {
            Class.forName("android.os.Build");
            return true;
        } catch (ClassNotFoundException e) {
            return false;
        }
    }

    private static Path findInDir(Path dir) {
        if (!Files.isDirectory(dir)) {
            return null;
        }
        List<String> names;
        if (Platform.isMac()) {
            names = Arrays.asList("librobot_bus_c.dylib", "librobot_bus.dylib");
        } else if (Platform.isWindows()) {
            names = Arrays.asList("robot_bus_c.dll", "robot_bus.dll");
        } else {
            names = Arrays.asList("librobot_bus_c.so", "librobot_bus.so");
        }
        for (String name : names) {
            Path file = dir.resolve(name);
            if (Files.isRegularFile(file)) {
                return file;
            }
        }
        return null;
    }
}

File: bindings/kotlin/src/main/kotlin/org/indunet/robot/bus/NativeLoader.kt
package org.indunet.robot.bus

import com.sun.jna.Native
import com.sun.jna.Platform
import java.nio.file.Files
import java.nio.file.Path

/**
 * Resolves the shared C ABI library built from `bindings/cpp/native`
 * (`librobot_bus_c` / `robot_bus_c`).
 *
 * Search order:
 * 1. `ROBOT_BUS_NATIVE` / `-Drobot.bus.native` — full path to the shared library
 * 2. `ROBOT_BUS_NATIVE_DIR` / `-Drobot.bus.native.dir` — directory containing it
 * 3. Sibling release build: `bindings/cpp/native/target/release`
 * 4. System library name `robot_bus_c` (java.library.path / DYLD_LIBRARY_PATH)
 */
internal object NativeLoader {
    fun loadLibrary(): RobotBusC {
        val explicit = System.getenv("ROBOT_BUS_NATIVE") ?: System.getProperty("robot.bus.native")
        if (!explicit.isNullOrBlank()) {
            return Native.load(explicit, RobotBusC::class.java)
        }
        val dir = System.getenv("ROBOT_BUS_NATIVE_DIR") ?: System.getProperty("robot.bus.native.dir")
        val candidates = buildList {
            if (!dir.isNullOrBlank()) add(Path.of(dir))
            add(Path.of("").toAbsolutePath().resolve("../cpp/native/target/release").normalize())
            add(Path.of(System.getProperty("user.dir"), "bindings/cpp/native/target/release"))
            // When cwd is bindings/kotlin
            add(Path.of(System.getProperty("user.dir"), "../cpp/native/target/release").normalize())
        }
        for (candidate in candidates) {
            val file = findInDir(candidate) ?: continue
            return Native.load(file.toAbsolutePath().toString(), RobotBusC::class.java)
        }
        return Native.load("robot_bus_c", RobotBusC::class.java)
    }

    private fun findInDir(dir: Path): Path? {
        if (!Files.isDirectory(dir)) return null
        val names = when {
            Platform.isMac() -> listOf("librobot_bus_c.dylib", "librobot_bus.dylib")
            Platform.isWindows() -> listOf("robot_bus_c.dll", "robot_bus.dll")
            else -> listOf("librobot_bus_c.so", "librobot_bus.so")
        }
        return names.map { dir.resolve(it) }.firstOrNull { Files.isRegularFile(it) }
    }
}
